### ISSUE TITLE
ICMP checksum 0xFFFF is invalid (RFC 1624)

### ISSUE THREAD

**Issue by ValZapod** [the-tcpdump-group/tcpdump#855]
It looks like ICMP 1624 outright says that
 
> the checksum field can never contain ~(+0), which is -0 (0xFFFF). It can, however, contain ~(-0), which is +0 (0x0000)

Also see https://bugs.wireshark.org/bugzilla/show_bug.cgi?id=16334. In https://bugs.wireshark.org/bugzilla/show_bug.cgi?id=5770 there is an attachment "Checksum corner case, successfully tracked w/r45007." th second packet of which has 0xFFFF, I think tcpdump -r thatfile should warn about wrong checksum, that should be 0x0000 or whatever there is.

I also edited wikipedia about that.

**Comment by ValZapod** [comment]
@guyharris This should be `wrong icmp cksum fffc (->0000)`
```
root@kali# tcpdump -v  -r './icmp_csum_with_fffc_but_should.pcap'  
reading from file /home/ball/Документы/icmp_csum_with_fffc_but_should.pcap, link-type EN10MB (Ethernet)
05:15:23.644701 IP (tos 0x0, ttl 60, id 0, offset 0, flags [DF], proto ICMP (1), length 84)
    192.168.1.1 > 192.168.1.2: ICMP echo request, id 1835, seq 1, length 64 (wrong icmp cksum c7ff (->f7ff)!)
05:15:23.644731 IP (tos 0x0, ttl 128, id 11885, offset 0, flags [DF], proto ICMP (1), length 84)
    192.168.1.2 > 192.168.1.1: ICMP echo reply, id 1835, seq 1, length 64 (wrong icmp cksum fffc (->ffff)!)

```
[icmp_csum_with_fffc_but_should.zip](https://github.com/the-tcpdump-group/tcpdump/files/4743328/icmp_csum_with_fffc_but_should.zip)

**Comment by guyharris** [comment]
See [my comment in Wireshark bug 16334 about this](https://bugs.wireshark.org/bugzilla/show_bug.cgi?id=16334#c23).

tl;dr: if you compute the one's-complement sum of a carefully-constructed and valid ICMP Echo Reply packet, the sum most definitely *will* be zero, so the one's complement of that sum most definitely *will* be 0xFFFF.  I've seen implementations "in the wild" (well, one's in the wild, and the other one is in the OS of the machine sitting on my lap) that compute 0x0000 and that compute 0xFFFF.  Other 0xFFFF's are the result of incorrectly-done incremental header updates by intermediate hosts; checksum checkers should accept either one.  Sniffers should warn about 0xFFFF, but perhaps not report it as an *error*, as there's no MUST or MUST NOT about the ICMP checksum.

**Comment by ValZapod** [comment]
> most definitely _will_ be 0xFFFF

What it should be here?? Because if you will hex edit the file and edit to 0xFFFF there or 0x0000, it will print in both cases with -v (so it checks for the sum) that it is correct!
```
05:15:23.644731 IP (tos 0x0, ttl 128, id 11885, offset 0, flags [DF], proto ICMP (1), length 84)
    192.168.1.2 > 192.168.1.1: ICMP echo reply, id 1835, seq 1, length 64

```

**Comment by ValZapod** [comment]
So as I understand it, we should add a check if there exist non-zero bits IN ICMP itself (no psuedo header, not like in ICMPv6) && checksum calculated is 0xFFFF, the real checksum is 0x0000 (and recommend it and not 0xFFFF when the checksum is wrong).

**Comment by ValZapod** [comment]
TCP packet with 0xFFFF was also seen in the wild! Though it is TCP in IPV4 **in GPRS** https://gitlab.com/wireshark/wireshark/-/issues/16816

That caused a https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2020-25862 in Wireshark. Nice.

**Comment by infrastation** [comment]
From Guy's comment is seems as if there is no problem, is this the conclusion of this case?

**Comment by ValZapod** [comment]
> there is no problem, is this the conclusion of this case?

What? No, there is only NO problem in ICMV6, because even if all bits are zero due to no using those bits only in a psuedo header there are some non-zero bits and thus the bug is impossible. But there is still a bug in ICMPv4 that does not use psuedoheader and thus 0xFFFF is possible.

**Comment by infrastation** [comment]
Thank you for your comments. If there is still a problem, it would help to state it better.

In the specific case of IPv4 ICMP the checksum can be 0x0000, which appears in the wire encoding as 0xFFFF. That's what everyone agrees about, correct?

In the specific case of tcpdump, one function of a packet analyzer is to display precisely what is on the wire, does tcpdump do that correctly for the checksum?

Another function is to flag certain or potential anomalies in the packet data, including suggestions of values that the receiving protocol instance would likely be expecting instead of what is on the wire. Does tcpdump do that correctly for the checksum?

**Comment by ValZapod** [comment]
> In the specific case of IPv4 ICMP the checksum can be 0x0000, which appears in the wire encoding as 0xFFFF. That's what everyone agrees about, correct?

Not correct at all. It does not appear on the wire in any way different. Checksum 0x0000 and 0xFFFF are both possible in icmpv4, 0xffff is only possible when you have all zero bits. 0x0000 is possible much more often.
> does tcpdump do that correctly for the checksum?

Yes. It just reads bytes there.

>Does tcpdump do that correctly for the checksum?

No, it does not warn that the checksum 0xFFFF is wrong in the case nonzero bits are present.
